### ISSUE TITLE
False positive with strict-type-predicates reports using union generic type

### ISSUE THREAD

**Issue by vietthang** [palantir/tslint#3222]
### Bug Report

- __TSLint version__: 5.5.0
- __TypeScript version__: 2.5.2
- __Running TSLint via__: CLI

#### TypeScript code being linted

```ts
export type Wrap<T> = {
  value: T,
}

export function test<T extends string>(valueOrWrap: T | Wrap<T>): T {
  if (typeof valueOrWrap === 'string') {
    return valueOrWrap
  }
  return valueOrWrap.value
}
```

with `tslint.json` configuration:

```json
{
  "rules": {
    "strict-type-predicates": true
  }
}
```

#### Actual behavior
`testWrap.ts[6, 7]: Expression is always false.`
#### Expected behavior
No error.

**Comment by ajafff** [comment]
While investigating this bug I found `// TODO: could use checker.getBaseConstraintOfType to help with type parameters, but it's not publicly exposed.` in the code.

The rule ignored type parameters before. It just didn't ignore unions that contain type parameters.
I'll submit a PR shortly.

**Comment by ajafff** [comment]
Turns out we don't need `checker.getBaseConstraintOfType`.
That means we don't need to ignore type parameters at all.

**Comment by bschlenk** [comment]
This might be the same/similar issue. I have an interface that looks like this:

```ts
interface MyInterface<T> {
    onMount?: (element: T) => void | { current: T | null };
}
```

When I use it and check if `onMount` is a function (`if (typeof onMount === 'function')`) I get the error "Expression is always false."

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is deprecated** and no longer accepting pull requests other than security fixes. See #4534. ☠️ 
We recommend you instead use **[typescript-eslint](https://typescript-eslint.io)** to lint your TypeScript code with **ESLint**. ✅

👋 It was a pleasure open sourcing with you!

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈 _(#4534)_  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
